src/protocol.rs

```diff
@@ -641,81 +555,153 @@ mod tests {
+
+        // Set up keygen participants
+        let keygen_sid = Identifier::random(&mut rng);
+        let configs = ParticipantConfig::random_quorum(QUORUM_SIZE, &mut rng);
+        let mut keygen_quorum = configs
+            .into_iter()
+            .map(|config| Participant::<KeygenParticipant>::from_config(&config, keygen_sid, ()))
+            .collect::<Vec<_>>();
+        let mut inboxes = HashMap::from_iter(
+            keygen_quorum
+                .iter()
+                .map(|p| (p.id, vec![]))
+                .collect::<Vec<_>>(),
+        );
+        let mut keygen_outputs: HashMap<
+            ParticipantIdentifier,
+            <KeygenParticipant as PublicProtocolParticipant>::Output,
+        > = HashMap::new();
+
+        // Initialize keygen for all participants
+        for participant in &keygen_quorum {
             let inbox = inboxes.get_mut(&participant.id).unwrap();
-            inbox.push(message);
-        }
-        while !is_presigning_done(&quorum, presign_identifier)? {
-            process_messages(&mut quorum, &mut inboxes, &mut rng)?;
+            inbox.push(participant.initialize_message());
         }
 
-        // Now, produce a valid signature
-        let mut hasher = Sha256::new();
-        hasher.update(b"some test message");
+        // Run keygen until all parties have outputs
+        while keygen_outputs.len() < QUORUM_SIZE {
+            let output = process_messages(&mut keygen_quorum, &mut inboxes, &mut rng)?;
 
-        let mut aggregator = SignatureShare::default();
-        for participant in &mut quorum {
-            let signature_share = participant.sign(presign_identifier, hasher.clone())?;
-            aggregator = aggregator.chain(signature_share)?;
-        }
-        let signature = aggregator.finish()?;
-
-        // Initialize all participants and get their public keyshares to construct the
-        // final signature verification key
-        let mut vk_point = CurvePoint::IDENTITY;
-        for participant in &mut quorum {
-            let X = participant.get_public_keyshare(keyshare_identifier)?;
-            vk_point = CurvePoint(vk_point.0 + X.0);
+            if let Some((pid, output)) = output {
+                assert!(keygen_outputs.insert(pid, output).is_none());
+            }
         }
-        let verification_key =
-            k256::ecdsa::VerifyingKey::from_encoded_point(&vk_point.0.to_affine().into()).unwrap();
 
-        // Moment of truth, does the signature verify?
-        assert!(verification_key.verify_digest(hasher, &signature).is_ok());
+        drop(keygen_quorum);
+
+        // Set up presign participants
+        let _presign_sid = Identifier::random(&mut rng);
+        let _configs = ParticipantConfig::random_quorum(QUORUM_SIZE, &mut rng);
+
+        /*
+                // Prepare presign inputs: a pair of outputs from keygen and auxinfo, with the same participant ID for each one.
+                // Blocking on #236.
```

Comment:
note: Formatting input for presign is why this is blocking on #236. I guess we could put it all into a `Storage` but I think that is not a useful intermediate state.